From 4accf02a811ba5fbc661128e698075acdf996523 Mon Sep 17 00:00:00 2001
From: Andreas Baierl <ichgeh@imkreisrum.de>
Date: Fri, 18 Sep 2026 14:40:23 +0200
Subject: [PATCH] amlogic: disable all quirks

This works with a current amlogic-vdec stateless decoder, which was recently merged into LibreELEC.

Amlogic has some small issues still (green artifacts on stream start until the second keyframe ...)
---
 README.md        | 10 +++-------
 codec_video.cpp  | 16 ----------------
 codec_video.h    |  2 --
 hardwaredevice.h | 18 ++----------------
 videostream.cpp  | 11 +++--------
 5 files changed, 8 insertions(+), 49 deletions(-)

diff --git a/README.md b/README.md
index ceb5ff54..5ec41988 100644
--- a/README.md
+++ b/README.md
@@ -36,7 +36,7 @@ Current development happens on a **Raspberry PI 4** and a **Radxa Rock 4B Plus (
 |                   | 576i MPEG2 | 720p H.264 | 1080i H.264 | 1080p HEVC |
 | ----------------- | ---------- | ---------- | ----------- | ---------- |
 | Allwinner         | Not tested | Not tested | Not tested  | Not tested |
-| Amlogic           | Not tested | Not tested | Not tested  | Not tested |
+| Amlogic           | ✅         | ✅         | ✅          | ✅         |
 | Raspberry Pi 2    | ❌         | SW         | SW          | SW         |
 | Raspberry Pi 3    | Not tested | Not tested | Not tested  | Not tested |
 | Raspberry Pi 4    | SW         | ✅         | ✅          | ✅         |
@@ -103,7 +103,6 @@ and is updated on every commit automatically.
 
 Known Bugs/ TODO
 ----------------
-- amlogic trickspeed is broken
 - rpi avcodec_flush_buffers is broken in [rpi-ffmpeg](https://github.com/jc-kynesim/rpi-ffmpeg) (used in LibreELEC), use a workaround for now
 - see https://github.com/rellla/vdr-plugin-softhddevice-drm-gles/issues
 
@@ -427,7 +426,7 @@ specific patches.
 
 The following instructions may help you to setup FFmpeg (may be outdated):
 
-- Raspberry Pi, Amlogic:
+- Raspberry Pi
 	- [rpi-ffmpeg](https://github.com/jc-kynesim/rpi-ffmpeg) is the very recent version for RPI4/RPI5 and Amlogic (For Raspberry Pi LibreELEC normally patches the upstream FFmpeg version to get the version from jc-kynesim)
 	- check out a recent branch (note: the master branch is not the one you want to have)
 	- most likely this branch has everything you need and you don't need any further patches (in doubt, check [how it is handled in LibreELEC](https://github.com/LibreELEC/LibreELEC.tv/tree/master/packages/multimedia/ffmpeg))
@@ -435,11 +434,8 @@ The following instructions may help you to setup FFmpeg (may be outdated):
 	- For Raspberry Pi building with the following configure options should enable hardware acceleration:
 
 			--disable-static --enable-shared --enable-pic --enable-bsfs --enable-filters --enable-v4l2_m2m --enable-libdrm --enable-libudev --enable-v4l2-request --enable-sand --enable-hwaccels --enable-neon --disable-vdpau --disable-vaapi --disable-mmal
-	- For Amlogic building with the following configure options should enable hardware acceleration:
 
-			--disable-static --enable-shared --enable-pic --enable-bsfs --enable-filters --enable-v4l2_m2m --enable-libdrm --enable-hwaccels --enable-neon --disable-vdpau --disable-vaapi --disable-libudev --disable-v4l2-request
-
-- Rockchip, Allwinner:
+- Rockchip, Allwinner, Amlogic:
 	- use mainline FFmpeg with version mentioned in [LibreELEC package file](https://github.com/LibreELEC/LibreELEC.tv/blob/master/packages/multimedia/ffmpeg/package.mk)
 	- patch the mainline version with the [patchsets](https://github.com/LibreELEC/LibreELEC.tv/tree/master/packages/multimedia/ffmpeg/patches) which where mentioned for your platform in the package.mk
 	- build FFmpeg as usual with the "configure - make - make install" logic
diff --git a/codec_video.cpp b/codec_video.cpp
index 54564139..735104c3 100644
--- a/codec_video.cpp
+++ b/codec_video.cpp
@@ -503,22 +503,6 @@ int cVideoDecoder::ReceiveFrame(AVFrame **frame)
 	if (pFrame->flags == AV_FRAME_FLAG_CORRUPT)
 		LOGDEBUG2(L_CODEC, "videocodec: %s: %s: AV_FRAME_FLAG_CORRUPT", m_identifier, __FUNCTION__);
 
-	// Codec artifacts workaround for amlogic H264:
-	// Skip m_skipKeyFramesNum Key-Frames at stream start.
-	// m_skipKeyFramesNum can be set with SetSkipKeyFramesNum()
-	if (m_pVideoCtx->codec_id == AV_CODEC_ID_H264 && m_skipKeyFramesNum && m_cntStartKeyFrames) {
-		if (IsKeyFrame(pFrame)) {
-			LOGDEBUG2(L_CODEC, "videocodec: %s: %s: artifact workaround - skip %s Keyframe nr %d", m_identifier, __FUNCTION__,
-				isInterlacedFrame(pFrame) ? "interlaced" : "progressive", m_cntStartKeyFrames);
-
-			if (m_cntStartKeyFrames++ > m_skipKeyFramesNum - 1)
-				m_cntStartKeyFrames = 0;
-		}
-
-		av_frame_free(&pFrame);
-		return AVERROR(EAGAIN);
-	}
-
 	*frame = pFrame;
 
 	if (!m_cntFramesReceived) {
diff --git a/codec_video.h b/codec_video.h
index 58087a53..a563656e 100644
--- a/codec_video.h
+++ b/codec_video.h
@@ -48,7 +48,6 @@ public:
 	const char *Name(void) { return m_pCodecString; };
 	int GetPacketsSent(void) { return m_cntPacketsSent; };
 	int GetFramesReceived(void) { return m_cntFramesReceived; };
-	void SetSkipKeyFramesNum(int num) { m_skipKeyFramesNum = num; };
 
 private:
 	AVCodecContext *m_pVideoCtx = nullptr;  ///< video codec context
@@ -62,7 +61,6 @@ private:
 	                                        ///< in ReceiveFrame() before sending them to the renderer)
 	int m_lastCodedWidth;                   ///< save coded width while closing for a directly reopen
 	int m_lastCodedHeight;                  ///< save coded height while closing for a directly reopen
-	int m_skipKeyFramesNum = 0;             ///< number of Keyframes (= I-Frames in VDR) to be skipped at stream start (hardware specific quirk)
 	bool m_isHardwareDecoder = false;       ///< true, if this is a hardware decoder
 
 	int GetExtraData(const AVPacket *);
diff --git a/hardwaredevice.h b/hardwaredevice.h
index 6a483fed..83edc591 100644
--- a/hardwaredevice.h
+++ b/hardwaredevice.h
@@ -25,11 +25,7 @@
  * @ingroup misc
  */
 enum HardwareQuirks {
-	QUIRK_NO_HW_DEINT                 = 1 << 0,     ///< set, if no hw deinterlacer available
-	QUIRK_CODEC_FLUSH_WORKAROUND      = 1 << 1,     ///< set, if we have to close and reopen the codec instead of avcodec_flush_buffers (rpi)
-	QUIRK_CODEC_NEEDS_DIMENSION_PARSE = 1 << 2,     ///< set, if codec needs some infos for init (coded_width and coded_height)
-	QUIRK_CODEC_SKIP_FIRST_FRAMES     = 1 << 3,     ///< set, if codec should skip first I-Frames
-	QUIRK_CODEC_SKIP_NUM_FRAMES       = 2     ,     ///< skip QUIRK_CODEC_SKIP_NUM_FRAMES, in case QUIRK_CODEC_SKIP_FIRST_FRAMES is set
+	QUIRK_CODEC_FLUSH_WORKAROUND       = 1 << 0,     ///< set, if we have to close and reopen the codec instead of avcodec_flush_buffers (rpi)
 };
 
 /**
@@ -120,13 +116,6 @@ public:
 				m_quirks |= QUIRK_CODEC_FLUSH_WORKAROUND;
 				break;
 			}
-			if (strstr(read_ptr, "amlogic")) {
-				m_deviceName = "amlogic";
-				m_quirks |= QUIRK_CODEC_NEEDS_DIMENSION_PARSE
-				         |  QUIRK_CODEC_SKIP_FIRST_FRAMES
-				         |  QUIRK_NO_HW_DEINT;
-				break;
-			}
 
 			read_size -= strlen(read_ptr) + 1;
 			read_ptr = (char *)&read_ptr[(strlen(read_ptr) + 1)];
@@ -134,10 +123,7 @@ public:
 
 		if (m_deviceName)
 			LOGDEBUG("%s found%s%s%s%s", m_deviceName,
-			    m_quirks & QUIRK_NO_HW_DEINT ?                 ", hw deinterlacer disabled" : "",
-			    m_quirks & QUIRK_CODEC_FLUSH_WORKAROUND ?      ", flush workaround" : "",
-			    m_quirks & QUIRK_CODEC_NEEDS_DIMENSION_PARSE ? ", parse H.264 dimensions" : "",
-			    m_quirks & QUIRK_CODEC_SKIP_FIRST_FRAMES ?     ", skip first I-Frames" : "");
+			    m_quirks & QUIRK_CODEC_FLUSH_WORKAROUND ?       ", flush workaround" : "");
 		else
 			LOGDEBUG("%s found, no quirks set", txt_buf);
 
diff --git a/videostream.cpp b/videostream.cpp
index 218389a8..37368603 100644
--- a/videostream.cpp
+++ b/videostream.cpp
@@ -144,8 +144,6 @@ void cVideoStream::StartDecoder()
 	LOGDEBUG2(L_CODEC, "videostream %s: %s", m_identifier, __FUNCTION__);
 
 	m_pDecoder = new cVideoDecoder(m_identifier);
-	if (m_hardwareQuirks & QUIRK_CODEC_SKIP_FIRST_FRAMES)
-		m_pDecoder->SetSkipKeyFramesNum(QUIRK_CODEC_SKIP_NUM_FRAMES);
 
 	Start();
 }
@@ -221,9 +219,7 @@ void cVideoStream::OpenDecoder(void)
 	int width = 0;
 	int height = 0;
 
-	bool needsParsing = m_startDecodingWithIFrame ||
-	                    m_parseH264Dimensions ||
-	                   (m_hardwareQuirks & QUIRK_CODEC_NEEDS_DIMENSION_PARSE);
+	bool needsParsing = m_startDecodingWithIFrame || m_parseH264Dimensions;
 
 	if (needsParsing && m_codecId == AV_CODEC_ID_H264) {
 		cH264Parser h264Packet(m_packets.Peek(),
@@ -244,8 +240,7 @@ void cVideoStream::OpenDecoder(void)
 			return;
 		}
 
-		// amlogic h264 decoder needs width an height for correct decoder open
-		if ((m_hardwareQuirks & QUIRK_CODEC_NEEDS_DIMENSION_PARSE) || m_parseH264Dimensions) {
+		if (m_parseH264Dimensions) {
 			width = h264Packet.GetWidth();
 			height = h264Packet.GetHeight();
 			LOGDEBUG2(L_CODEC, "videostream %s: %s: Parsed width %d height %d", m_identifier, __FUNCTION__, width, height);
@@ -621,7 +616,7 @@ void cVideoStream::RenderFrame(AVFrame * frame)
 		// - AV_PIX_FMT_YUV420P, progressive -> scale filter to get NV12 frames
 		// - AV_PIX_FMT_DRM_PRIME, interlaced, hw deinterlacer available -> hw deinterlacer
 		if (frame->format == AV_PIX_FMT_YUV420P ||
-		   (frame->format == AV_PIX_FMT_DRM_PRIME && m_useDeinterlacer && (!(m_hardwareQuirks & QUIRK_NO_HW_DEINT))))
+		   (frame->format == AV_PIX_FMT_DRM_PRIME && m_useDeinterlacer))
 			m_videoFilter.InitAndStart(m_pDecoder->GetContext(), frame, m_useDeinterlacer);
 
 		m_checkFilterThreadNeeded = false;
-- 
2.47.3

